### PR TITLE
Added id/meta/terms endpoint.

### PR DESCRIPTION
**Pull request by alesstimec**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1407/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1408/
Test PASSed.

**Comment by mhilton** [comment]
:+1:

File: docs/API.md [hunk 1]
```diff
@@ -1073,6 +1073,20 @@ Example: `GET trusty/juju-gui-3/meta/charm-config`
 }
 ```
 
+#### GET *id*/meta/terms
+
+The `meta/terms` path returns a list of terms and angreements the user must
```
Comment (rogpeppe) [inline docs/API.md]: a/angreements/agreements/

Comment (rogpeppe) [inline docs/API.md]: ... but don't you mean "conditions" here?

File: docs/API.md [hunk 2]
```diff
@@ -1073,6 +1073,20 @@ Example: `GET trusty/juju-gui-3/meta/charm-config`
...
+agree to in order to obtain the archive of the given charm or bundle id.
+
+Example: `GET some-charm/meta/terms`
+
+```json
+[
+    "enterprise-terms/1",
+    "special-terms/17",
```
Comment (rogpeppe) [inline docs/API.md]: The final comma is not legal JSON here.

File: docs/API.md [hunk 3]
```diff
@@ -1073,6 +1073,20 @@ Example: `GET trusty/juju-gui-3/meta/charm-config`
 }
 ```
 
+#### GET *id*/meta/terms
+
+The `meta/terms` path returns a list of terms and angreements the user must
+agree to in order to obtain the archive of the given charm or bundle id.
```
Comment (rogpeppe) [inline docs/API.md]: Perhaps document that this endpoint comes directly from the terms field in the charm or bundle metadata?

Comment (rogpeppe) [inline docs/API.md]: In fact, bundles can't have terms, so "bundle id" is wrong here AFAICS. In fact, is it actually worth having a separate terms endpoint when we can just fetch the charm metadata and get the terms from that?

File: internal/v5/api.go [hunk 4]
```diff
@@ -601,6 +602,15 @@ func (h *ReqHandler) metaCharmConfig(entity *mongodoc.Entity, id *router.Resolve
 	return entity.CharmConfig, nil
 }
 
+// GET id/meta/terms
+// https://github.com/juju/charmstore/blob/v4/docs/API.md#get-idmetaterms
+func (h *ReqHandler) metaTerms(entity *mongodoc.Entity, id *router.ResolvedURL, path string, flags url.Values, req *http.Request) (interface{}, error) {
+	if entity.CharmMeta == nil || len(entity.CharmMeta.Terms) == 0 {
```
Comment (rogpeppe) [inline internal/v5/api.go:611]: I think this should probably return nil, nil when the entity is not a charm (which will result in an ErrMetadataNotFound error, same as when any charm-specific endpoint is accessed in a bundle). Or is the intention that terms might apply to bundles in the future too?

File: internal/v5/api_test.go [hunk 5]
```diff
@@ -816,6 +834,25 @@ func (s *APISuite) TestMetaPermPutUnauthorized(c *gc.C) {
 	})
 }
 
+func (s *APISuite) TestMetaTerms(c *gc.C) {
```
Comment (rogpeppe) [inline internal/v5/api_test.go:840]: Please add a test for retrieving terms for a bundle too (to test whatever behaviour is decided upon)

**Comment by rogpeppe** [comment]
Seems reasonable in itself, but I have a few comments and queries about this.

File: docs/API.md (hunk 1 above)
Comment (alesstimec) [inline docs/API.md]: indeed, conditions.. done

File: docs/API.md (hunk 2 above)
Comment (alesstimec) [inline docs/API.md]: done.

File: docs/API.md (hunk 3 above)
Comment (alesstimec) [inline docs/API.md]: that is true, but if i recall correctly, @urosj requested this endpoint..

File: internal/v5/api.go (hunk 4 above)
Comment (alesstimec) [inline internal/v5/api.go:611]: done.

File: internal/v5/api_test.go (hunk 5 above)
Comment (alesstimec) [inline internal/v5/api_test.go:840]: done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1409/
Test PASSed.

**Comment by rogpeppe** [comment]
LGTM thanks!

**Comment by alesstimec** [comment]
$$merge$$

**Comment by alesstimec** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**